src/lualdap.c

```diff
@@ -438,17 +450,27 @@ static int result_message (lua_State *L) {
 	}
 }
 
+static future_data* prepare_future(lua_State *L, int conn_idx, int code) {
+	future_data *future = lua_newuserdata(L, sizeof(future_data*));
+	lua_createtable(L, 0, 1);
+	lua_pushvalue(L, conn_idx);
+	lua_setfield(L, -2, "connection");
+	lua_setuservalue(L, -2);
+	luaL_setmetatable(L, LUALDAP_FUTURE_METATABLE);
+	future->code = code;
+	return future;
+}
```

Comment:
So that you can `prepare` before you `return`.
It fixes the scenario where allocation of the return object fails (which would otherwise result in leaking the connection object)

Reply:
I think this belongs into one PR with the other changes to how LuaLDAP uses futures, right?

Reply:
Only the final commit to change `:bind_simple()` is API breaking.
Feel free to merge this PR only up to the point you're comfortable with. I generally order commits from most acceptable to most contentious.